ci: add shellcheck linter for shell scripts

ggilmore opened the pull request:
# Overview 

This PR adds the popular https://github.com/koalaman/shellcheck to our CI pipeline. Shellcheck helps prevent a lot of shell footguns and pitfalls and will make our scripts more-robust overall. 

I fixed all of these scripts by hand following shellcheck's suggestions. I'm reasonably confident that the critical paths work (CI pipelines + docker builds pass, dev/start.sh). There might be some lesser-used scripts that are slightly broken by these changes, but I'm confident that those be fixed up after they're discovered by follow up PRs.


# Riskier edits
I tagged people here so that they can sign off on the more extensive / riskier script fixes that I had to make. 

**I'd like for these to signed off on before merging**. 


- [x] @sourcegraph/core-services https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/ci/ci-db-backcompat.sh (Note: https://github.com/sourcegraph/sourcegraph/issues/9921 will be fixed in a separate PR)
- [x] @sourcegraph/core-services https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/ci/db-backcompat.sh (Note: https://github.com/sourcegraph/sourcegraph/issues/9921 will be fixed in a separate PR)
- [x] @sourcegraph/core-services https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/go-install.sh
- [x] @sourcegraph/core-services https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/handle-change.sh
- [x] @efritz https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/squash_migrations.sh
- [x] @sourcegraph/distribution @ryan-blunden https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/src-expose/entry.sh (**Note**: user-facing script)
- [x] @sourcegraph/distribution @uwedeportivo https://github.com/sourcegraph/sourcegraph/blob/shellcheck/docker-images/prometheus/entry.sh (**Note**: user-facing script / https://github.com/koalaman/shellcheck/wiki/SC2039)


# Scripts sorted by code owner

This PR touches a lot of scripts. This section conveniently lists all of the edited files group by their code owner so that each team can take a look at the scripts that they're supposed to maintain. 

Note: I'm not asking each team to review every one of the files here before I merge this PR (the high-risk files I flagged are listed above). Most of the changes here are fairly straightforward. This section is just an easy way to navigate the PR. I'm confident that issues that slip through here can be easily fixed in a follow up PR. 


@sourcegraph/web 
* [browser/scripts/release-ff.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/browser/scripts/release-ff.sh)
* [dev/licenses-npm.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/licenses-npm.sh)

@sourcegraph/nobody 
* [dev/Procfile](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/Procfile)	@sourcegraph/nobody
* [dev/add_migration.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/add_migration.sh)	@sourcegraph/nobody
* [dev/api/delete_user.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/api/delete_user.sh)	@sourcegraph/nobody
* [dev/api/delete_user_mutation.json](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/api/delete_user_mutation.json)	@sourcegraph/nobody
* [dev/api/user_id_query.json](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/api/user_id_query.json)	@sourcegraph/nobody
* [dev/auth-provider/keycloak.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/auth-provider/keycloak.sh)	@sourcegraph/nobody
* [dev/auth-provider/scripts/common.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/auth-provider/scripts/common.sh)	@sourcegraph/nobody
* [dev/auth-provider/scripts/configure-keycloak.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/auth-provider/scripts/configure-keycloak.sh)	@sourcegraph/nobody
* [dev/caddy.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/caddy.sh)	@sourcegraph/nobody
* [dev/changewatch.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/changewatch.sh)	@sourcegraph/nobody
* [dev/check/all.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/check/all.sh)	@sourcegraph/nobody
* [dev/check/bash-syntax.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/check/bash-syntax.sh)	@sourcegraph/nobody
* [dev/check/broken-urls.bash](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/check/broken-urls.bash)	@sourcegraph/nobody
* [dev/check/check-owners.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/check/check-owners.sh)	@sourcegraph/nobody
* [dev/check/docsite.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/check/docsite.sh)	@sourcegraph/nobody
* [dev/check/go-dbconn-import.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/check/go-dbconn-import.sh)	@sourcegraph/nobody
* [dev/check/go-enterprise-import.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/check/go-enterprise-import.sh)	@sourcegraph/nobody
* [dev/check/go-lint.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/check/go-lint.sh)	@sourcegraph/nobody
* [dev/check/gofmt.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/check/gofmt.sh)	@sourcegraph/nobody
* [dev/check/licenses.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/check/licenses.sh)	@sourcegraph/nobody
* [dev/check/no-localhost-guard.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/check/no-localhost-guard.sh)	@sourcegraph/nobody
* [dev/check/shellcheck.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/check/shellcheck.sh)	@sourcegraph/nobody
* [dev/check/yarn-deduplicate.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/check/yarn-deduplicate.sh)	@sourcegraph/nobody
* [dev/ci/ci-db-backcompat.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/ci/ci-db-backcompat.sh)	@sourcegraph/nobody
* [dev/ci/db-backcompat.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/ci/db-backcompat.sh)	@sourcegraph/nobody
* [dev/ci/e2e.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/ci/e2e.sh)	@sourcegraph/nobody
* [dev/ci/parallel_run.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/ci/parallel_run.sh)	@sourcegraph/nobody
* [dev/ci/reset-test-db.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/ci/reset-test-db.sh)	@sourcegraph/nobody
* [dev/ci/yarn-build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/ci/yarn-build.sh)	@sourcegraph/nobody
* [dev/ci/yarn-run.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/ci/yarn-run.sh)	@sourcegraph/nobody
* [dev/ci/yarn-test-separate.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/ci/yarn-test-separate.sh)	@sourcegraph/nobody
* [dev/ci/yarn-test.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/ci/yarn-test.sh)	@sourcegraph/nobody
* [dev/comby-install-or-upgrade.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/comby-install-or-upgrade.sh)	@sourcegraph/nobody
* [dev/dev-sourcegraph-server.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/dev-sourcegraph-server.sh)	@sourcegraph/nobody
* [dev/foreach-ts-project.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/foreach-ts-project.sh)	@sourcegraph/nobody
* [dev/generate.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/generate.sh)	@sourcegraph/nobody
* [dev/git-diff-no-enterprise.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/git-diff-no-enterprise.sh)	@sourcegraph/nobody
* [dev/go-install.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/go-install.sh)	@sourcegraph/nobody
* [dev/handle-change.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/handle-change.sh)	@sourcegraph/nobody
* [dev/jaeger.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/jaeger.sh)	@sourcegraph/nobody
* [dev/owners](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/owners)	@sourcegraph/nobody
* [dev/owners.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/owners.sh)	@sourcegraph/nobody
* [dev/phabricator/e2e.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/phabricator/e2e.sh)	@sourcegraph/nobody
* [dev/phabricator/install-sourcegraph.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/phabricator/install-sourcegraph.sh)	@sourcegraph/nobody
* [dev/phabricator/restart.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/phabricator/restart.sh)	@sourcegraph/nobody
* [dev/phabricator/shared.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/phabricator/shared.sh)	@sourcegraph/nobody
* [dev/phabricator/start.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/phabricator/start.sh)	@sourcegraph/nobody
* [dev/postgres_exporter.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/postgres_exporter.sh)	@sourcegraph/nobody
* [dev/prune-pick.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/prune-pick.sh)	@sourcegraph/nobody
* [dev/run-server-image.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/run-server-image.sh)	@sourcegraph/nobody
* [dev/start.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/start.sh)	@sourcegraph/nobody
* [dev/syntect_server](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/syntect_server)	@sourcegraph/nobody
* [dev/syntect_server.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/syntect_server.sh)	@sourcegraph/nobody
* [enterprise/dev/dev-sourcegraph-server.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/enterprise/dev/dev-sourcegraph-server.sh)	@sourcegraph/nobody
* [enterprise/dev/generate.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/enterprise/dev/generate.sh)	@sourcegraph/nobody
* [enterprise/dev/start.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/enterprise/dev/start.sh)	@sourcegraph/nobody

@sourcegraph/distribution 
* [.buildkite/hooks/pre-command](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/.buildkite/hooks/pre-command)	@sourcegraph/distribution @ggilmore
* [.tool-versions](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/.tool-versions)	@sourcegraph/distribution
* [cmd/frontend/build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/cmd/frontend/build.sh)	@sourcegraph/distribution
* [cmd/github-proxy/build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/cmd/github-proxy/build.sh)	@sourcegraph/distribution
* [cmd/gitserver/build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/cmd/gitserver/build.sh)	@sourcegraph/distribution
* [cmd/loadtest/build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/cmd/loadtest/build.sh)	@sourcegraph/distribution
* [cmd/query-runner/build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/cmd/query-runner/build.sh)	@sourcegraph/distribution
* [cmd/replacer/build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/cmd/replacer/build.sh)	@sourcegraph/distribution
* [cmd/repo-updater/build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/cmd/repo-updater/build.sh)	@sourcegraph/distribution
* [cmd/searcher/build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/cmd/searcher/build.sh)	@sourcegraph/distribution
* [cmd/server/build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/cmd/server/build.sh)	@sourcegraph/distribution
* [cmd/server/jaeger.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/cmd/server/jaeger.sh)	@sourcegraph/distribution
* [cmd/server/pre-build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/cmd/server/pre-build.sh)	@sourcegraph/distribution
* [dev/check/build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/check/build.sh)	@sourcegraph/distribution
* [dev/grafana.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/grafana.sh)	@sourcegraph/distribution
* [dev/libsqlite3-pcre/build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/libsqlite3-pcre/build.sh)	@sourcegraph/distribution
* [dev/prometheus.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/prometheus.sh)	@sourcegraph/distribution
* [dev/src-expose/build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/src-expose/build.sh)	@sourcegraph/distribution
* [docker-images/alpine/build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/docker-images/alpine/build.sh)	@sourcegraph/distribution
* [docker-images/alpine/release.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/docker-images/alpine/release.sh)	@sourcegraph/distribution
* [docker-images/grafana/build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/docker-images/grafana/build.sh)	@sourcegraph/distribution
* [docker-images/indexed-searcher/build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/docker-images/indexed-searcher/build.sh)	@sourcegraph/distribution
* [docker-images/jaeger-agent/build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/docker-images/jaeger-agent/build.sh)	@sourcegraph/distribution
* [docker-images/jaeger-all-in-one/build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/docker-images/jaeger-all-in-one/build.sh)	@sourcegraph/distribution
* [docker-images/postgres-11.4/build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/docker-images/postgres-11.4/build.sh)	@sourcegraph/distribution
* [docker-images/prometheus/build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/docker-images/prometheus/build.sh)	@sourcegraph/distribution
* [docker-images/prometheus/entry.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/docker-images/prometheus/entry.sh)	@sourcegraph/distribution
* [docker-images/redis-cache/build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/docker-images/redis-cache/build.sh)	@sourcegraph/distribution
* [docker-images/redis-store/build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/docker-images/redis-store/build.sh)	@sourcegraph/distribution
* [docker-images/search-indexer/build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/docker-images/search-indexer/build.sh)	@sourcegraph/distribution
* [docker-images/syntax-highlighter/build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/docker-images/syntax-highlighter/build.sh)	@sourcegraph/distribution
* [enterprise/cmd/frontend/build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/enterprise/cmd/frontend/build.sh)	@sourcegraph/distribution
* [enterprise/cmd/repo-updater/build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/enterprise/cmd/repo-updater/build.sh)	@sourcegraph/distribution
* [enterprise/cmd/server/build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/enterprise/cmd/server/build.sh)	@sourcegraph/distribution

@sourcegraph/core-services 
* [enterprise/dev/check/build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/enterprise/dev/check/build.sh)	@sourcegraph/distribution
@sourcegraph/core-services 
* [cmd/symbols/go-build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/cmd/symbols/go-build.sh)	@sourcegraph/core-services

@slimsag 
* [cmd/frontend/pre-build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/cmd/frontend/pre-build.sh)	@slimsag
* [enterprise/cmd/frontend/pre-build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/enterprise/cmd/frontend/pre-build.sh)	@slimsag
* [enterprise/cmd/server/pre-build.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/enterprise/cmd/server/pre-build.sh)	@slimsag

@keegancsmith 
* [dev/src-expose/docker-publish.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/src-expose/docker-publish.sh)	@keegancsmith
* [dev/src-expose/entry.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/src-expose/entry.sh)	@keegancsmith

@felixfbecker 
* [.vscode/extensions.json](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/.vscode/extensions.json)	@felixfbecker

@efritz 
* [dev/squash_migrations.sh](https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/squash_migrations.sh)	@efritz

keegancsmith commented:
@ggilmore pushed 175d396

Note some of the advice actually breaks stuff. For example in that change I had to escape \$ because moving to `$(cat` from the multiline bash strings `'` meant bash would interpret some strings.

ggilmore commented:
> @ggilmore pushed [175d396](https://github.com/sourcegraph/sourcegraph/commit/175d396bf617e44c891fe626cac968385221bef4)
> 
> Note some of the advice actually breaks stuff. For example in that change I had to escape $ because moving to `$(cat` from the multiline bash strings `'` meant bash would interpret some strings.

Thanks! This was my fault though. Note that https://github.com/koalaman/shellcheck/wiki/SC2016 doesn't actually say that this is a problem in every case. This was just me being overzealous and using a heredoc because I thought it'd be easier to read.

ggilmore commented:
> do we document that we support this? I added this, but since then I've moved to using direnv. Not sure if anyone is using it, and if they are they should use direnv.

@keegancsmith I'd prefer addressing this in another PR instead of cluttering up this one.

unknwon commented:
Thanks for the improvement!

I see a big list of files owned by nobody, may not be important to this PR, but who should review and own them?

ggilmore commented:
@unknwon In the long run, a fair number of those files should probably be owned by @sourcegraph/distribution or @sourcegraph/core-services since a lot of them deal with our dev environment. However, at the moment @sourcegraph/nobody is legitimately the owner of these files: https://github.com/sourcegraph/sourcegraph/blob/master/.github/CODEOWNERS#L11

I have done my best to go through and flag the edits that I thought were more involved in the "riskier edits" section. I have specifically requested reviewers to look at those files

It's possible that there is something that I missed, but I'm reasonably confident that the scripts that aren't specifically called out are 1)) not in the critical path path scripts or 2) any issues will be discovered fixed up after this PR is merged. Feel free to look at them anyway though.

slimsag commented:
> @slimsag: https://github.com/sourcegraph/sourcegraph/blob/shellcheck/dev/api/delete_user.sh

Delete this script. It has been superseded by the `src` CLI.

slimsag commented:
@ggilmore It's going to be very difficult for you to land this PR as-is with the number of reviews you are asking for. Additionally, when you tag @sourcegraph/distribution it's not really clear who on that team should be reviewing this.

I would suggest you split up this PR into three categories:

1. Low-risk edits that anybody on the team can review and approve.
2. Medium-risk edits that anybody on the team can review and approve.
2. High-risk edits that need review from a single specific code owner.

slimsag commented:
Ah, never-mind, I misread and thought you were asking for review on everything under "Scripts sorted by code owner"

ggilmore commented:
> @ggilmore It's going to be very difficult for you to land this PR as-is with the number of reviews you are asking for. Additionally, when you tag @sourcegraph/distribution it's not really clear who on that team should be reviewing this.
> 
> I would suggest you split up this PR into three categories:
> 
>     1. Low-risk edits that anybody on the team can review and approve.
> 
>     2. Medium-risk edits that anybody on the team can review and approve.
> 
>     3. High-risk edits that need review from a single specific code owner.

@slimsag 
I went ahead and clarified that I'm not blocked on teams reviewing every single file change in that list.

uwedeportivo commented:
this is very educational. very humbling :-).

unknwon commented:
> @unknwon In the long run, a fair number of those files should probably be owned by @sourcegraph/distribution or @sourcegraph/core-services since a lot of them deal with our dev environment. However, at the moment @sourcegraph/nobody is their legitimate owner: [/.github/CODEOWNERS@`master`#L11](https://github.com/sourcegraph/sourcegraph/blob/master/.github/CODEOWNERS#L11)

Agree @sourcegraph/distribution and @sourcegraph/core-services are the nature owners of most of these files. And @sourcegraph/nobody serves the exact purpose that it should never appear as the code owner as its team description states. We can deal with it later, issue filed: https://github.com/sourcegraph/sourcegraph/issues/10064

unknwon commented:
I can't confident reviewing bash scripts, could @keegancsmith or @mrnugget help review four files need to be signed off under **Riskier edits** section?

mrnugget commented on dev/ci/db-backcompat.sh:
```diff
@@ -64,12 +67,13 @@ function runTest() {
     # find which packages' tests need to be run.
-    PACKAGES_WITH_DB_TEST=$(go list -f '{{$printed := false}}{{range .TestImports}}{{if and (not $printed) (eq . "github.com/sourcegraph/sourcegraph/cmd/frontend/db/testing")}}{{$.ImportPath}}{{$printed = true}}{{end}}{{end}}' ./...)
+    # shellcheck disable=SC2016
+    mapfile -t PACKAGES_WITH_DB_TEST < <(go list -f '{{$printed := false}}{{range .TestImports}}{{if and (not $printed) (eq . "github.com/sourcegraph/sourcegraph/cmd/frontend/db/testing")}}{{$.ImportPath}}{{$printed = true}}{{end}}{{end}}' ./...)
```
What's the reasoning behind this change here? Asking because the new version seems a bit more complicated.

ggilmore replied:
It was part of addressing https://github.com/koalaman/shellcheck/wiki/SC2086

Loading the output of this command into an explicit array lets us have full control over how we expand the array (avoiding word-splitting). 

The change on Line 76 is also part of this fix.

mrnugget commented on dev/ci/e2e.sh:
```diff
@@ -28,11 +28,10 @@ socat tcp-listen:7080,reuseaddr,fork system:"docker exec -i $CONTAINER socat std
 echo "--- Waiting for $URL to be up"
 set +e
-timeout 60s bash -c "until curl --output /dev/null --silent --head --fail $URL; do
+if ! timeout 60s bash -c "until curl --output /dev/null --silent --head --fail $URL; do
```
nit: I'd personally pull the `"until curl [...]"` out into a variable. That makes it easier to read that this is a multiline condition.

mrnugget commented on dev/handle-change.sh:
```diff
@@ -34,11 +36,11 @@ for i; do
       cmd=${i#cmd/}
       cmd=${cmd%%/*}
-      case " $cmdlist " in
+      case " ${cmdlist[*]} " in
```
I've stared at these three lines for 5min now and all I can think is: "Bash is ridiculous."

For anyone reading along, what this code does is to check whether the `cmd` already exists in `cmdlist`:

```bash
bash-5.0$ cmdlist=()
bash-5.0$ cmd="gitserver"
bash-5.0$ case " ${cmdlist[*]} " in         " $cmd ") ;;         *)           cmdlist+=("$cmd");           ;;       esac
bash-5.0$ echo "${cmdlist[@]}"
gitserver
bash-5.0$ case " ${cmdlist[*]} " in         " $cmd ") ;;         *)           cmdlist+=("$cmd");           ;;       esac
bash-5.0$ echo "${cmdlist[@]}"
gitserver
```

The problem, and I think this was already a problem in the existing code, is that the check only works if there's a single element in `cmdlist`:

```bash
bash-5.0$ cmdlist=("gitserver" "frontend")
bash-5.0$ case " ${cmdlist[*]} " in         " $cmd ") ;;         *)           cmdlist+=("$cmd");           ;;       esac
bash-5.0$ echo "${cmdlist[@]}"
gitserver frontend gitserver
bash-5.0$ case " ${cmdlist[*]} " in         " $cmd ") ;;         *)           cmdlist+=("$cmd");           ;;       esac
bash-5.0$ echo "${cmdlist[@]}"
gitserver frontend gitserver gitserver
```

Again, this might be pre-existing behaviour, I just stumbled on it while following along.

ggilmore replied:
I also realized this was an issue while I was working on it, but I didn't want to alter the existing behavior of the script.

ggilmore commented:
> Signing off on those 4 files, but take my approval with a grain of salt: that Bash code is full of mystery to me. I learned a ton while following along, but I still think we should replace these scripts with something >Bash.

I don't disagree (using something like Go for re-implementing stuff like dev/start.sh). I still think shellcheck is worth having so that it can warn us about issues like this.
